pilot/pkg/model/policyattachment.go

```diff
@@ -0,0 +1,66 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"istio.io/api/type/v1beta1"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+type policyTargetGetter interface {
+	GetTargetRef() *v1beta1.PolicyTargetReference
+	GetSelector() *v1beta1.WorkloadSelector
+}
+
+type policyMatch string
+
+const (
+	policyMatchSelector policyMatch = "selector"
+	policyMatchDirect   policyMatch = "direct"
+	policyMatchIgnore   policyMatch = "ignore"
+)
+
+func getPolicyMatcher(policyName string, opts workloadSelectionOpts, policy policyTargetGetter) policyMatch {
+	gatewayName, isGatewayAPI := opts.workloadLabels[constants.IstioGatewayLabel]
+	targetRef := policy.GetTargetRef()
+	if isGatewayAPI && targetRef == nil && policy.GetSelector() != nil {
```

Comment:
remove policy.GetSelector() != nil

Reply:
For namespace or root namespace policies, we still want those to apply to waypoints and gateways. We just don't want selector policies to apply

Reply:
? I could not understand such a hack requirement.

Reply:
Then i would argue why canot we apply policies with selector

Reply:
The key is the different personas. If a platform admin sets a policy on the root mesh namespace, a namespace owner who deploys a waypoint for a particular service account should inherit that policy by default unless they write a conflicting one. This is how Istio works today, right?

Reply:
I see no difference with selector

Reply:
The difference is purely semantic. With a targetRef, it is impossible to write a policy that applies to a request/connection _after_ load balancing (which is undesirable and bloats the config considerably due to Envoy implementation). With workload selector, you can write such policies. While this is ok in sidecars, it's problematic in Ambient which is why targetRef as an API change prevents it. You can find more info in the [Ambient APIs doc](https://docs.google.com/document/d/1Lk1I1qB-XwVWFP0sVcIN5JMJZHjuF3j8lGWtUeDK6pY/edit#heading=h.8nuwse76cdev)

Reply:
Here I see waypoint as a special workload, just like ingress gateway(k8s). Can we disallow bind a policy with workload selector to a ingress gateway? I think no, we cannot.

As a user,  I don't care much about whether the policy is applied before or after loadbalancing. I am not clear what the problem it is for `This makes policies bound to workload selectors problematic.`

Reply:
The only reason workloadSelector worked as an API in the first place is because of the sidecar model: there's always a server sidecar to apply the policy after load balancing. To implement the same thing for Gateway API gateways (including waypoints) requires ugly internal listener logic in Envoy. Instead of carrying that forward, this initiative aligns with Gateway API by explicitly binding policy to the waypoints of the server

Reply:
I still do not get the persuasive point about this.

@howardjohn @linsun Can you explain the limitation why workload selector is a problem for gateway https://docs.google.com/document/d/1Lk1I1qB-XwVWFP0sVcIN5JMJZHjuF3j8lGWtUeDK6pY/edit

Reply:
The selector vs targetRef is not fundamental or _required_, it is slightly better in a variety of ways that add up to make it worthwhile (IMO).

* Selector is a bit implicit. We have pod, service, gateway, and policy all doing label selectors. They are all tied together, but its all very vague and you can have N:M:X:Y relationship between them which is pretty crazy to understand. TargetRef is much more explicit and cleaner to understand (similar to how we made Gateway:Service binding direct with Gateway API, too).
* TargetRef is the recommended solution for attaching policy to gateways, which is what we are doing. Aside from aligning with other products, there is also plans for tooling, conformance testing, etc around this.

The 'load balancing' discussion is about whether we apply a policy with `app=some-workload` to waypoint. In ambient initial launch, we _did_. In the latter launches, we changed the label selector to be on the gateway (e.g. `istio.io/gateway-name=my-waypoint`). This is a big difference. This decision is in many ways orthogonal from targetRef vs selector, but there is some correlation: the distinction that "workload selector" policies apply to workloads and "target ref" apply to Gateways may help guide users to correctly attaching these policies to the waypoint instead of their workloads. I don't think this is a compelling argument on its own, though.

tl;dr: targetRef is a better API model than workloadSelector and the ecosystem at large is moving towards it. This is simply moving us in that direction. It is not strictly required, more of an ergonomic improvement. The reason its important to do now is that ambient is alpha, so we can make breaking changes more easily.

Reply:
OK, got it. `more of an ergonomic improvement` makes sense to me.

As to `Namespace scope or root scoped policy can apply to gateway/waypoint `, where can we document it.

Reply:
How about in [this section](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#resource-attachment-and-scaling)? I also need to write up a more detailed doc on gw api